File: internal/layout/layout_test.go
package layout

import (
	"reflect"
	"testing"
)

func TestAssignmentStableSlots(t *testing.T) {
	var a Assignment
	got := a.Update([]string{"w1", "w2", "w3"})
	if !reflect.DeepEqual(got, []string{"w1", "w2", "w3"}) {
		t.Fatalf("initial fill: %v", got)
	}
	got = a.Update([]string{"w9", "w2", "w3"})
	if !reflect.DeepEqual(got, []string{"w2", "w3", "w9"}) {
		t.Fatalf("survivors must compact before additions append: %v", got)
	}
	got = a.Update([]string{"w9", "w2", "w3"})
	if !reflect.DeepEqual(got, []string{"w2", "w3", "w9"}) {
		t.Fatalf("unchanged input must not shuffle: %v", got)
	}
}

func TestAssignmentRemovalCompacts(t *testing.T) {
	var a Assignment
	a.Update([]string{"w1", "w2", "w3", "w4"})
	got := a.Update([]string{"w1", "w4"})
	if !reflect.DeepEqual(got, []string{"w1", "w4"}) {
		t.Fatalf("removals must compact: %v", got)
	}
}

func TestPagerActivation(t *testing.T) {
	if PagerActive(14, 14) {
		t.Fatal("14 spaces on 14 keys needs no pager")
	}
	if !PagerActive(15, 14) {
		t.Fatal("15 spaces on 14 keys needs a pager")
	}
	if PagerActive(3, 2) {
		t.Fatal("two keys cannot host the two-key pager")
	}
}

func TestPaging(t *testing.T) {
	keys := 14
	spaces := 15
	active := PagerActive(spaces, keys)
	size := PageSize(keys, active)
	if size != 12 {
		t.Fatalf("page size with two-key pager: %d", size)
	}
	if count := PageCount(spaces, size); count != 2 {
		t.Fatalf("page count: %d", count)
	}
	if start, end := VisibleRange(1, size, spaces); start != 12 || end != 15 {
		t.Fatalf("second page range: [%d, %d)", start, end)
	}
	if PageOf(14, size) != 1 {
		t.Fatalf("space 14 must live on page 1")
	}
}

func TestClamp(t *testing.T) {
	if Clamp(-1, 3) != 0 || Clamp(5, 3) != 2 || Clamp(1, 3) != 1 {
		t.Fatal("clamp bounds")
	}
}

File: test/layout.test.ts
import { describe, expect, it } from "vitest";

import { Assignment, clamp, pageCount, pageOf, pagerActive, pageSize, visibleRange } from "../src/layout";

describe("Assignment", () => {
	it("keeps survivors in stable slots", () => {
		const a = new Assignment();
		expect(a.update(["w1", "w2", "w3"])).toEqual(["w1", "w2", "w3"]);
		expect(a.update(["w9", "w2", "w3"])).toEqual(["w2", "w3", "w9"]);
		expect(a.update(["w9", "w2", "w3"])).toEqual(["w2", "w3", "w9"]);
	});

	it("compacts removals", () => {
		const a = new Assignment();
		a.update(["w1", "w2", "w3", "w4"]);
		expect(a.update(["w1", "w4"])).toEqual(["w1", "w4"]);
	});
});

describe("pager", () => {
	it("activates only when spaces exceed the keys", () => {
		expect(pagerActive(14, 14)).toBe(false);
		expect(pagerActive(15, 14)).toBe(true);
		expect(pagerActive(3, 2)).toBe(false);
	});

	it("pages with the two-key pager", () => {
		const keys = 14;
		const spaces = 15;
		const size = pageSize(keys, pagerActive(spaces, keys));
		expect(size).toBe(12);
		expect(pageCount(spaces, size)).toBe(2);
		expect(visibleRange(1, size, spaces)).toEqual([12, 15]);
		expect(pageOf(14, size)).toBe(1);
	});

	it("clamps the page", () => {
		expect(clamp(-1, 3)).toBe(0);
		expect(clamp(5, 3)).toBe(2);
		expect(clamp(1, 3)).toBe(1);
	});
});
